demo/Views/ApplicationView.vala

```diff
@@ -0,0 +1,81 @@
+/*
+ *  Copyright 2019 elementary, Inc. (https://elementary.io)
+ *
+ *  This program or library is free software; you can redistribute it
+ *  and/or modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 3 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General
+ *  Public License along with this library; if not, write to the
+ *  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301 USA.
+ */
+
+public class ApplicationView : Gtk.Grid {
+    construct {
+        var progress_visible_label = new Gtk.Label ("Show Progress:");
+        var progress_visible_switch = new Gtk.Switch ();
+        var progress_scale = new Gtk.Scale.with_range (Gtk.Orientation.HORIZONTAL, 0, 1, 0.05);
+        var badge_visible_label = new Gtk.Label ("Show Badge:");
+        var badge_visible_switch = new Gtk.Switch ();
+        var badge_spin = new Gtk.SpinButton.with_range (0, int64.MAX, 1);
+
+        column_spacing = 12;
+        row_spacing = 6;
+        orientation = Gtk.Orientation.VERTICAL;
+        halign = Gtk.Align.CENTER;
+        valign = Gtk.Align.CENTER;
+        attach (progress_visible_label, 0, 0, 1, 1);
+        attach (progress_visible_switch, 1, 0, 1, 1);
+        attach (progress_scale, 0, 1, 2, 1);
+        attach (badge_visible_label, 0, 2, 1, 1);
+        attach (badge_visible_switch, 1, 2, 1, 1);
+        attach (badge_spin, 0, 3, 2, 1);
+
+        progress_scale.value_changed.connect (() => {
+            Granite.Services.Application.set_progress.begin (progress_scale.get_value (), (obj, res) => {
+                try {
+                    Granite.Services.Application.set_progress.end (res);
```

Comment:
I don't think we need to have the async callback here at all. I believe Vala will automatically complete the callback with just calling `begin`.

Reply:
we want the callback because I'm using the error that might be throwed

Reply:
Okay then.